docs/spec_draft.md

```diff
@@ -1718,6 +1719,75 @@ and produces a `result` tensor. More formally,
 ```
 
 [Back to Ops](#index-of-ops)
+
+## stablehlo.rng_bit_generator
+
+### Semantics
+
+Returns an `output` filled with uniform random bits and an updated output state
+`output_state` given an initial state `initial_state` using the pseudorandom
+number generator algorithm `rng_algorithm`. The output is guaranteed to be
+deterministic function of `initial_state`, but it is not guaranteed to be
+deterministic between backends and different compiler versions.
+
+`initial_state` is the initial state of the current random number generation.
+The initial state, required shape, and valid values are dependent on the
+algorithm used.
+
+`rng_algorithm` is one of the following, different variants of the algorithm is
+implementation-defined:
+  * `DEFAULT`: Backend specific algorithm.
+  * `THREE_FRY`: Threefry counter-based PRNG algorithm.*
+  * `PHILOX`: Philox algorithm to generate random numbers in parallel.*
+
+\* See: [Salmon et al. SC 2011. Parallel random numbers: as easy as 1, 2, 3.
+](http://www.thesalmons.org/john/random123/papers/random123sc11.pdf)
+
+### Inputs
+
+| Name            | Type                                         |
+|-----------------|----------------------------------------------|
+| `initial_state` | tensor of integer or floating-point type     |
+| `rng_algorithm` | enum of `DEFAULT`, `THREE_FRY`, and `PHILOX` |
+
+### Outputs
+
+| Name           | Type                                     |
+|----------------|------------------------------------------|
+| `output_state` | tensor of integer or floating-point type |
+| `output`       | tensor of integer or floating-point type |
+
+### Constraints
+
+  * (C1) shape(`initial_state`) $=$ shape(`output_state`).
+  * (C2) The relationship between `initial_state`, `output_state`, and `output`
+      varies:
+    * If `rng_algorithm = DEFAULT`:
+      * `initial_state`, `output_state`, and `output` have backend-specific
+        requirements.
+    * If `rng_algorithm = THREE_FRY`:
+      * type(`initial_state`) $=$ type(`output_state`) $=$ `2xui64`.
+      * element_type(`output`) $\in$ {`ui32`, `ui64`}.
```

Comment:
I know we have already gone through this, but still have a few questions:
Is the second constraint of the `output` on element-type ever mentioned in the paper. I know that you did exhaustive experiments to find the behavior of HLO. The thing that I am afraid about is that what if the element-type(`output`), just like the shape('output`), is also dependent on the algorithm. 

From HLO spec
```
It and the required shape and valid values are dependent on the algorithm used. 
```
it does not say that the element-type is algo dependent, but OTOH we does not say what it is.

Can we investigate if this is just an implementation detail and not related to algorithm? If that is the case, then skip any constraint  on  the `output`.

Reply:
The output should be either `ui32` or `ui64` based on the variant of the algorithm used. As mentioned above, it is generating random bits not random numbers, so 32-bit algorithm would generate random 32-bits (similarly for 64 bits). As for whether 32-input can generate 64-bit output is not possible, since that is not how the algorithm is designed. It is possible, however, to use a 64-bit algorithm to generate a 32-bit value (and throwing away the other half), which is implementation-defined variant of the algorithm as stated in the spec.

Reply:
However, DEFAULT algorithm can do whatever it supports.

Reply:
Thanks for explaining.